pkg/cmd/pulumi/convert.go

```diff
@@ -222,15 +223,48 @@ func runConvert(
 				return result.FromError(fmt.Errorf("could not load pcl program: %w", err))
 			}
 		} else {
-			return result.FromError(fmt.Errorf("unrecognized source %s", from))
+			return result.FromError(fmt.Errorf("unrecognized source %q", from))
 		}
 	} else if from == "tf" {
 		proj, program, err = tfgen.Eject(cwd, loader, mapper)
 		if err != nil {
 			return result.FromError(fmt.Errorf("could not load terraform program: %w", err))
 		}
 	} else {
-		return result.FromError(fmt.Errorf("unrecognized source %s", from))
+		// Try and load the converter plugin for this
+		converter, err := plugin.NewConverter(pCtx, from, nil)
+		if err != nil {
+			return result.FromError(fmt.Errorf("plugin source %q: %w", from, err))
+		}
+
+		targetDirectory, err := os.MkdirTemp("", "pulumi-convert")
+		if err != nil {
+			return result.FromError(fmt.Errorf("create temporary directory: %w", err))
+		}
+
+		logging.Warningf("Plugin converters are currently experimental")
+
+		_, err = converter.ConvertProgram(pCtx.Request(), &plugin.ConvertProgramRequest{
+			SourceDirectory: cwd,
+			TargetDirectory: targetDirectory,
+		})
```

Comment:
converter is constructed using pCtx, any reason that it cannot hold onto the reference of pCtx.Request, to call it itself? It could remove a parameter and simplify usage.

```
		_, err = converter.ConvertProgram(pCtx.Request(), &plugin.ConvertProgramRequest{
			SourceDirectory: cwd,
			TargetDirectory: targetDirectory,
		})
```
```
		_, err = converter.ConvertProgram(&plugin.ConvertProgramRequest{
			SourceDirectory: cwd,
			TargetDirectory: targetDirectory,
		})
```

Reply:
> any reason that it cannot hold onto the reference of pCtx.Request, to call it itself? It could remove a parameter and simplify usage.

I think pCtx.Request is a design flaw and I didn't want to repeat it in this new code. Having a context per call means you can control the cancellation and context data for an individual call. E.g. For the Converter object on the server side it gets the grpc context as it's ctx parameter.